compute_tract_distance.py

```diff
@@ -0,0 +1,140 @@
+import argparse
+import vtk
+import numpy as np
+
+# This will be the mean closest fiber distance
+# Gerig et al. (2004) as ‘‘the mean of closest distances’’
+# It is defined as ‘‘the mean of the closest distance for every point of two fibers’’
+# 
+# Go through all points of fiber fm and find the point on fiber ft that is closest
+# Measure this distance and record it
+# Compute the average of all these distances
+def fiber_distance(points_m, points_t):
+  
+  # Reshape for vectorized calculation
+  points_m = points_m.reshape((len(points_m), 1, 3))
+  points_t = points_t.reshape((1, len(points_t), 3))
+
+  # fm - ft (num_points_f, num_points_t, 3) -- for an element i,j this is the coordinate distance between fm_i and ft_j
+  fm_minus_ft = points_m - points_t
+  squared = fm_minus_ft**2
+  
+  # We should sum 3rd dim (x,y,z) coords and take the sqrt
+  sum_of_squares = np.sum(squared, axis=2)
+  distances_fm_ft = np.sqrt(sum_of_squares)
+
+  # Now we the need the minimum value of each row which should represent for point fm_i the closest point on ft
+  mins_fm = np.min(distances_fm_ft, axis=1)
+
+  return np.mean(mins_fm)
+
+def main(fiber_bundle_M_path, fiber_bundle_T_path, use_every_n_fibers):
+
+  # Read fiber bundle M as polydata
+  reader_M = vtk.vtkPolyDataReader()
+  reader_M.SetFileName(fiber_bundle_M_path)
+  reader_M.Update()
+  fiber_bundle_M = reader_M.GetOutput()
+
+  # Read fiber bundle T as polydata
+  reader_T = vtk.vtkPolyDataReader()
+  reader_T.SetFileName(fiber_bundle_T_path)
+  reader_T.Update()
+  fiber_bundle_T = reader_T.GetOutput()
+
+  # These are all the points which make up the fibers (streamlines) of fiber bundle M
+  fiber_bundle_M_points = fiber_bundle_M.GetPoints()
+  fiber_bundle_M.GetLines().InitTraversal()
+  idListM = vtk.vtkIdList()
+
+  # These are all the points which make up fiber bundle T
+  fiber_bundle_T_points = fiber_bundle_T.GetPoints()
+
+  iter_counter_M = -1  
+  index_counter_M = 0
+
+  # Figure out how many fibers in M and T we will use based on the use_every_n_fiber value (needed for the normalization term)
+  num_fibers_M = int(np.ceil(fiber_bundle_M.GetNumberOfLines()/use_every_n_fibers))
+  num_fibers_T = int(np.ceil(fiber_bundle_T.GetNumberOfLines()/use_every_n_fibers))
+
+  # This is a matrix which stores all pairwise distances between fibers in M and T -- non-symmetric since d(fm, ft) != d(ft, fm)
+  # We cannot store the d in a shared matrix because d() is not symmetric so the diagonal doesn't work in this case
+  pairwise_distances_M_and_T = np.zeros((num_fibers_M, num_fibers_T))
+  pairwise_distances_T_and_M = np.zeros((num_fibers_T, num_fibers_M))
+
+  # Loop over the fiber bundle M
+  while (fiber_bundle_M.GetLines().GetNextCell(idListM)):
+   
+    iter_counter_M+=1
+
+    if (iter_counter_M % use_every_n_fibers != 0):
```

Comment:
Instead of literally using every $n^{\rm th}$ fiber I think we should randomly _sample_ a certain number of fibers. Then the parameter to specify, instead of `use_every_n_fibers` would be something more like `number_of_fibers_to_sample`. A random seed can be set to make everything deterministic/reproducible.

This would avoid having to worry about any unwanted coincidence of how the fibers are ordered in the vtk object and give a better guarantee that the fibers are being sampled randomly

Reply:
I have made this change and wrote a script to test the variability w.r.t. sampling. Here are the slides of a few experiments that we discussed. 
[fiber_tract_variability.pptx](https://github.com/brain-microstructure-exploration-tools/abcd-registration-experiments/files/14920254/fiber_tract_variability.pptx)